src/Disks/IDiskRemote.cpp

```diff
@@ -25,8 +25,41 @@ namespace ErrorCodes
     extern const int FILE_ALREADY_EXISTS;
     extern const int PATH_ACCESS_DENIED;;
     extern const int CANNOT_DELETE_DIRECTORY;
+    extern const int CANNOT_OPEN_FILE;
+    extern const int CANNOT_UNLINK;
+    extern const int CANNOT_FSTAT;
+    extern const int CANNOT_STAT;
 }
 
+namespace
+{
+
+// TODO: Remote object leaking is still possible when we fail to remove remote objects after
+// unlinking the last meta file.
+size_t removeFileAndReturnNumOfHardlinks(const char * path)
+{
+    int fd = ::open(path, 0);
+    if (fd == -1)
+    {
+        if (errno == ENOENT)
+            return 0;
+        throwFromErrno(fmt::format("Cannot open file: {}", path), ErrorCodes::CANNOT_OPEN_FILE);
+    }
+    if (-1 == ::unlink(path))
```

Comment:
btw recently all metadata related disk operations were moved to `metadata_disk` object. I think better to use it here too. May be we can just use here `metadata_disk->removeFile()` and `metadata_disk->getRefCount()` (not implemented for diskLocal now, but just add implementation with `fstat`?)

Reply:
The `removeFile` and `getRefCount` have to be under the same mutex along with `createHardLink`, which is the whole idea of this PR. And the method only makes sense for metadata_disk. I've renamed it to `removeMetaFileAndReturnNumOfHardlinks`. IDiskRemote should be a reasonable place for it.